libcfnet/net.c

```diff
@@ -225,6 +225,90 @@ int ReceiveTransaction(const ConnectionInfo *conn_info, char *buffer, int *more)
     return ret;
 }
 
+/* BWlimit global variables
+
+  Throttling happens for all network interfaces, all traffic being sent for
+  any connection of this process (cf-agent or cf-serverd).
+  We need a lock, to avoid concurrent writes to "bwlimit_next".
+  Then, "bwlimit_next" is the absolute time (as of clock_gettime() ) that we
+  are clear to send, after. It is incremented with the delay for every packet
+  scheduled for sending. Thus, integer arithmetic will make sure we wait for
+  the correct amount of time, in total.
+ */
+
+static pthread_mutex_t bwlimit_lock = PTHREAD_MUTEX_INITIALIZER;
+static struct timespec bwlimit_next = {0, 0L};
+uint32_t bwlimit_kbytes = 0; /* desired limit, in kB/s */ 
+
+
+/** Throttle traffic, if next packet happens too soon after the previous one
+ * 
+ *  This function is global, accross all network operations (and interfaces, perhaps)
+ *  @param tosend Length of current packet being sent out (in bytes)
+ */
+
+void EnforceBwLimit(int tosend)
+{
+    const uint32_t u_10e6 = 1000000L;
+    const uint32_t u_10e9 = 1000000000L;
+    struct timespec clock_now;
+
+    if (!bwlimit_kbytes)
+    {
+        /* early return, before any expensive syscalls */
+        return;
+    }
+
+    if (pthread_mutex_lock(&bwlimit_lock) == 0)
+    {
+        clock_gettime(CLOCK_MONOTONIC, &clock_now);
```

Comment:
It appears that ``CLOCK_MONOTONIC`` is Linux-specific. I propose that you use ``CLOCK_REALTIME`` instead and store the old value, so you can detect backwards jumps and reset the next time. If the clock is adjusted there may be small breaks or bursts, but this effect will be temporary.